UWP Improvements

basharast opened the pull request:
Hi,

Here a suggested changes to improve the UWP version

changes were carefully added to avoid breaking other platforms, it will be applied only for UWP

I don't know what is used exactly in libretro core, so all changes excluded from libretro build


## Main changes
- Custom memory stick location 
- Launch options (File, URI)
- UWP storage ~layer~ manager

these changes may address: #15595  & #16423

## Custom memory stick

I used the same UI in Android to make user able to change the location to another place

I did my best to enable this feature without breaking anything else.

![MStick](https://user-images.githubusercontent.com/3244951/235243728-317b0504-ec7c-4886-8c40-4f262d33cf5f.PNG)


## Launch options

- User can now launch PPSSPP (UWP) using file
- or using URI `ppsspp:` with support to switch back to the caller
`ppsspp:?cmd="fullpath"` or `ppsspp:?cmd="fullpath"&launchOnExit=targetURI`


## ~UWP storage layer~

Major part of this layer pulled out due to not required by latest windows builds, only builds before 10.0.17134


## Features enabled

I enabled some features to make it almost similar to the desktop version such as browse folder

now it will generate logs ~near to~ inside PSP folder ~inside LOGS folder~ (not sure if this was enabled before but in case it wasn't)


## Need to more work [Solved]

~Some places still can be better, like the way to detect launch request~

~I didn't find a better and effective place other than Render (loop) to keep track these requests~

~it need to be moved to proper place in future~.


I hope you will find it easy to check and helpful

already tested on XBOX, Desktop without any problem.

Thanks.

basharast committed:
UWP Improved Support

hrydgard commented:
This is really cool, thanks for doing this!

I'll review it in a bit more detail tomorrow, and will probably end up merging it after the 1.15 release process (we don't yet do official UWP releases anyway).

basharast committed:
minor fixes

hrydgard commented on Common/File/DirListing.cpp:
```diff
@@ -67,6 +71,11 @@ bool GetFileInfo(const Path &path, FileInfo * fileInfo) {
 	if (!GetFileAttributesExFromAppW(path.ToWString().c_str(), GetFileExInfoStandard, &attrs)) {
+#if !defined(__LIBRETRO__)
+	if (GetFileInfoUWP(path.ToString(), fileInfo)) {
+			return true;
```
nitpick: double deep indentation

hrydgard commented on Common/File/FileUtil.cpp:
```diff
@@ -157,7 +160,18 @@ FILE *OpenCFile(const Path &path, const char *mode) {
 
 #if defined(_WIN32) && defined(UNICODE)
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
+	// We shouldn't use _wfopen specificly for custom memory stick location
```
"specifically". Also, maybe can write a quick line about the reason.

hrydgard commented on Common/File/FileUtil.cpp:
```diff
@@ -258,6 +272,10 @@ static bool ResolvePathVista(const std::wstring &path, wchar_t *buf, DWORD bufSi
 		HANDLE hFile = CreateFile2FromAppW(path.c_str(), GENERIC_READ, FILE_SHARE_READ, OPEN_EXISTING, nullptr);
+#if !defined(__LIBRETRO__)
+		//Use UWP StorageManager to get handle
+		hFile = CreateFileUWP(path, GENERIC_READ, FILE_SHARE_READ, OPEN_EXISTING);
```
So in this case, we first call CreateFile2FromAppW and then CreateFileUWP? Doesn't seem quite right, the CreateFile2FromAppW call should probably be moved into an #else here.

hrydgard commented on UWP/PackageNormal.appxmanifest:
```diff
@@ -23,6 +23,21 @@
         </uap:InitialRotationPreference>
       </uap:VisualElements>
+	  <Extensions>
+        <uap:Extension Category="windows.fileTypeAssociation">
+          <uap:FileTypeAssociation Name="psprom">
+            <uap:SupportedFileTypes>
+              <uap:FileType>.cso</uap:FileType>
+              <uap:FileType>.bin</uap:FileType>
+              <uap:FileType>.iso</uap:FileType>
+              <uap:FileType>.elf</uap:FileType>
+            </uap:SupportedFileTypes>
+          </uap:FileTypeAssociation>
+        </uap:Extension>
```
not sure what happened to indentation here, but maybe doesn't really matter.

hrydgard commented on UWP/UWPHelpers/StorageExtensions.cpp:
```diff
@@ -0,0 +1,316 @@
+	return ret;
+}
+
+std::wstring make_wstring(const std::string& string)
```
we have utilities for this kind of stuff already, in Common/Data/Encoding/Utf8.h . I guess it can still make sense to have it here if you want to make the StorageManager code reusable between apps.

hrydgard commented on UWP/UWPHelpers/StorageExtensions.cpp:
```diff
@@ -0,0 +1,316 @@
+	return cpy;
+}
+
+std::vector<std::string> split(const std::string s, char seperator)
```
Similarly, we have this in Common/StringUtils.h.

hrydgard commented on UWP/UWPHelpers/StorageExtensions.h:
```diff
@@ -0,0 +1,93 @@
+// Telegram: @basharastifan
+// GitHub: https://github.com/basharast/UWP2Win32
+
+// Functions:
+// replace(std::string& str, const std::string& from, const std::string& to)
+// replace2(const std::string str, const std::string& from, const std::string& to)
+// split(const std::string s, char seperator)
+// 
+// isChild(std::string parent, std::string child)
+// isParent(std::string parent, std::string child, std::string childName)
```
I don't really understand why you repeat the function declarations in comments - they'll just quickly drift out of date when changes are made.

hrydgard commented on UWP/UWPHelpers/StorageFileW.h:
```diff
@@ -0,0 +1,281 @@
+		return storageFile;
+	}
+
+    time_t  filetime_to_timet(LARGE_INTEGER ull) const {
+		return ull.QuadPart / 10000000ULL - 11644473600ULL;
```
indentation problem

hrydgard reviewed (changes requested):
First round of comments!

Also, I'm wondering if all those libretro checks is needed in callers - maybe it's possible to have the storagemanager use dummy implementations in libretro mode, or something.

basharast commented:
Thanks man, I really appreciate your time to check this PR, 
applying the changes now, will submit them later today.
as for (string functions/helpers) I'm in very early stage in C/C++ it wasn't easy to learn that and understand the project structure in short time,
so I forgot to look in the code for possible matches, I noticed that later after I finished, I will replace these as well.
and yeah libretro check better to be moved to the storage manager instead of the current case.

basharast committed:
Major update

basharast commented:
So apparently I had few things mixed up due to the old builds of windows and lack of information in MS docs
We actually don't need to go through the slow UWP async functions
once the folder/file is added to the future access list then:

## API direct access
it will be allowed for direct access using the API that made for UWP (those ends with `FromApp`)
this is allowed even without FileSystem enabled.

## Get StorageFile with full path
also functions such as `StorageFolder::GetFolderFromPathAsync` allowed to get files from outside local data
as long the target in the future access list

## PR
I had to pull out many changes, let me know if I should re-submit another pull and close this in case this going to cause confusing.

GABO1423 commented:
> ## Code formating
> for some reason in VS lines look aligned and good, but in github some of them appear not aligned.
> 
The good old tabs vs spaces indentation issue, in GitHub a tab is equal to around 8 spaces (IIRC) while in Visual Studio a tab is much less than that. It's kind of annoying to deal with but you get used to it once you know about this.

unknownbrackets commented on UI/NativeApp.cpp:
```diff
@@ -756,7 +786,11 @@ void NativeInit(int argc, const char *argv[], const char *savegame_dir, const ch
 
 	DEBUG_LOG(SYSTEM, "ScreenManager!");
 	g_screenManager = new ScreenManager();
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
+	if (g_Config.memStickDirectory.empty() || (g_Config.bFirstRun && isLocalState(g_Config.memStickDirectory.ToString()))) {
```
Since UWP controls this (`g_Config.memStickDirectory = Path(memstickFolderW);`), why not just set it to blank there if this case matches?

I guess the intention here is to always show this prompt on first run, right?

-[Unknown]

basharast replied:
Yes, I think if `g_Config.bFirstRun` is enough here, I don't remember why I made the other checks

unknownbrackets commented on UWP/App.cpp:
```diff
@@ -14,8 +14,11 @@
 #include <ppltasks.h>
 
-using namespace UWP;
+#include <UWPHelpers/LaunchItem.h>
```
Please use `""` around files that are local to the project, and `<>` around system files.  This is our file.

-[Unknown]

unknownbrackets commented on UWP/App.cpp:
```diff
@@ -235,9 +252,10 @@ void App::OnSuspending(Platform::Object^ sender, SuspendingEventArgs^ args) {
 		app->m_deviceResources->Trim();
-		deferral->Complete();
-	});
+	    deferral->Complete();
```
Spaces and tabs mixed here.  I'm not sure what IDE you're using, but look for a plugin that enables EditorConfig.  It'll make it easier not to use the wrong indentation style across projects - even when they have different rules.

Also, please prefer:
```c++
create_task([app, deferral]() {
	asdf();
	bar();
});
```

Over:

```c++
create_task([app, deferral]() {
	asdf();
	bar();
	});
```

-[Unknown]

unknownbrackets commented on UWP/App.cpp:
```diff
@@ -277,7 +295,8 @@ void App::OnVisibilityChanged(CoreWindow^ sender, VisibilityChangedEventArgs^ ar
 		// Try to save the config here
-		g_Config.Save("App::OnVisibilityChanged");
+		// OnSuspending should now do the job
+		//g_Config.Save("App::OnVisibilityChanged");
```
If this isn't needed, I recommend removing the entire if.  Git will remember.

-[Unknown]

unknownbrackets commented on UWP/PPSSPP_UWPMain.cpp:
```diff
@@ -37,6 +37,12 @@
+// UWP Storage helper includes
+#include "UWPHelpers/StorageManager.h"
+#include "UWPHelpers/StorageAsync.h"
+#include <UWPHelpers/LaunchItem.h>
```
nit: Should be `""` style include.

-[Unknown]

unknownbrackets commented on UWP/PPSSPP_UWPMain.cpp:
```diff
@@ -367,12 +392,15 @@ std::vector<std::string> System_GetPropertyStringVec(SystemProperty prop) {
 		tempPath.resize(sz);
 		result.push_back(ConvertWStringToUTF8(tempPath));
 
-		if (getenv("TMPDIR") && strlen(getenv("TMPDIR")) != 0)
+		// As per below 'getenv' isn't supported
+		// https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/getenv-wgetenv?view=msvc-170
+		/*if (getenv("TMPDIR") && strlen(getenv("TMPDIR")) != 0)
```
FWIW, this is used to tell a user there's a problem when their savedata is being stored in a temp folder (this happens if you launch the non-UWP build from within a zip file) since your savedata will get lost between runs.

We can just remove these lines for UWP and rely only on `GetTempPath()`.  It should be fine.

-[Unknown]

unknownbrackets commented on UWP/PPSSPP_UWPMain.cpp:
```diff
@@ -462,48 +501,95 @@ void System_Notify(SystemNotification notification) {
+	}
+	case SystemRequestType::BROWSE_FOR_FILE:
+	{
+		std::vector<std::string> supportedExtensions = { ".cso", ".bin", ".iso", ".elf", ".zip" };
```
Probably PBP as well if we have ELF.

-[Unknown]

unknownbrackets commented on UWP/PPSSPP_UWPMain.cpp:
```diff
@@ -462,48 +501,95 @@ void System_Notify(SystemNotification notification) {
+			break;
+		case BrowseFileType::ANY:
+			supportedExtensions = {};
+			break;
```
I recommend using a `default:` here.  For defensive programming, the best would be:

```
switch ((BrowseFileType)param3) {
...
case BrowseFileType::BOOTABLE:
	break;
default:
	ERROR_LOG(FILESYS, "Unexpected BrowseFileType: %d", param3);
	return false;
}
```

This is called "defensive programming."  It's a technique where you assume other parts of the code will change, and the person who changes them will not realize your code also needs to change.  By logging and returning false, it becomes more likely someone will figure out the cause of the error instead of making them spend hours debugging why something doesn't work, not realizing it's because the case is missing here.

It's almost a certainty that this will eventually change in PPSSPP and the person who changes it will very likely forget to update UWP.

-[Unknown]

unknownbrackets replied:
Hm, wait, you didn't add `BrowseFileType::BOOTABLE`, so I think this will fail for that type of dialog currently.

-[Unknown]

basharast replied:
Oops, I shouldn't forget to add it back sorry about that

unknownbrackets commented on UWP/PPSSPP_UWPMain.cpp:
```diff
@@ -528,7 +614,13 @@ bool System_MakeRequest(SystemRequestType type, int requestId, const std::string
+		// Unsupported
+	}
+	else {
+		OpenFolder(std::string(path));
```
Technically, the provided path is to a file, and we should be opening its parent folder.  On non-UWP this prompts Explorer to open its parent folder and to highlight the specified file.  Is `OpenFolder()` going to be happy with a filename?

-[Unknown]

basharast replied:
Oh, then this need to be solved within `OpenFolder`

unknownbrackets commented on UWP/UWP.vcxproj:
```diff
@@ -89,12 +89,14 @@
     <AppxAutoIncrementPackageRevision>True</AppxAutoIncrementPackageRevision>
     <AppxSymbolPackageEnabled>False</AppxSymbolPackageEnabled>
     <AppxBundle>Never</AppxBundle>
-    <PackageCertificateThumbprint>C8DEB388B9BC89D1DC61324E4E9D9FE6A796B7AA</PackageCertificateThumbprint>
```
People keep changing this line, I haven't really involved myself in it.  Is this change intentional?

-[Unknown]

basharast replied:
No, I don't know actually why VS is changing this, since I never touched the certificate at all.

unknownbrackets commented on ext/libzip/zip_source_file_win32_utf16.c:
```diff
@@ -95,7 +119,8 @@ utf16_create_file(const char *name, DWORD access, DWORD share_mode, PSECURITY_AT
 
     return CreateFile2((const wchar_t *)name, access, share_mode, creation_disposition, &extParams);
 #else
-    return CreateFile2FromAppW((const wchar_t *)name, access, share_mode, creation_disposition, NULL);
+    HANDLE h = CreateFile2FromAppW((const wchar_t *)name, access, share_mode, creation_disposition, NULL);
+	return h;
```
This change doesn't look very necessary?  We should limit patches to ext/ code.

-[Unknown]

unknownbrackets commented on ext/libzip/zip_source_file_win32_utf16.c:
```diff
@@ -41,6 +41,29 @@ static HANDLE __stdcall utf16_create_file(const char *name, DWORD access, DWORD
 static void utf16_make_tempname(char *buf, size_t len, const char *name, zip_uint32_t i);
 static char *utf16_strdup(const char *string);
 
+#ifdef MS_UWP && !defined(__LIBRETRO__)
```
Why would we want to exclude LIBRETRO?

Also, you may want to propose these changes to https://github.com/nih-at/libzip.

-[Unknown]

unknownbrackets commented on UWP/UWPHelpers/StoragePickers.cpp:
```diff
@@ -0,0 +1,74 @@
+	    if (folder != nullptr)
+	    {
+			AddItemToFutureList(folder);
+	    	path = folder->Path;
```
Still some mixing of spaces and tabs in this file.  I recommend searching for "    " and cleaning up that way.  Should be quick.

-[Unknown]

basharast replied:
Sorry for asking this, but I'm a bit confused from the formating, in VS it look like below
![Capture](https://user-images.githubusercontent.com/3244951/235406992-cd4c5d52-cc7b-4541-b0af-2cc83ce11492.PNG)

is there any tool can help to fix these before submitting to GitHub?

unknownbrackets replied:
You can press Ctrl-R, then Ctrl-W to show whitespace indicators.  I usually search for a tab and it highlights where the indentation is funny.  But that said, most of the time I just don't have this problem, except maybe when pasting code from outside or something.

-[Unknown]

unknownbrackets commented on UWP/UWPHelpers/StorageManager.cpp:
```diff
@@ -0,0 +1,331 @@
+}
+
+bool isWriteMode(const char* mode) {
+	return (!strcmp(mode, "w") || !strcmp(mode, "wb") || !strcmp(mode, "wt") || !strcmp(mode, "at") || !strcmp(mode, "a"));
```
These are more or less "flags", so the order is not actually fixed.  I recommend instead using:

```c++
return strpbrk(mode, "wa+") != nullptr;
```

Which is roughly equivalent to the more obviously named:
```c++
return std::string(mode).first_first_of("wa") != std::string::npos;
```

unknownbrackets commented on UWP/UWPHelpers/StorageManager.cpp:
```diff
@@ -0,0 +1,331 @@
+	if (isWrite) {
+		access = GENERIC_WRITE;
+		share = FILE_SHARE_WRITE;
+		creation = isAppend ? OPEN_ALWAYS : CREATE_ALWAYS;
```
Actually, `"a"` means that all writes should be forced to the end of the file.

If the file was opened for "r+" (or rb+, or r+b, or rt+, or r+t), it would be writable but would not be created if it doesn't exist.  "w+" would create still.

-[Unknown]

basharast replied:
After the commit below it should be more accurate, I didn't know there are too many cases, hopefully all covered in the new function

unknownbrackets commented on UWP/UWPHelpers/StorageManager.cpp:
```diff
@@ -0,0 +1,331 @@
+				info.size = 1;
+				info.isDirectory = true;
+				info.isWritable = (attributes & FILE_ATTRIBUTE_READONLY) == 0;
+				info.atime = 1000;
```
0 might be a better "I don't know".

-[Unknown]

unknownbrackets replied:
Oh, I meant for the atime/mtime/ctime.  We might actually want isWritable to be true - this is only used by some games when they run `stat()`, but for that one `true` is probably safest if we don't know.

-[Unknown]

unknownbrackets commented on UWP/UWPHelpers/StorageAccess.cpp:
```diff
@@ -0,0 +1,117 @@
+#include "StorageAsync.h"
+#include "StorageAccess.h"
+#include "UWPUtil.h"
+#include <Common/File/Path.h>
```
nit: Should use `""` style include for our files.

-[Unknown]

basharast commented on UI/MainScreen.cpp:
```diff
@@ -1167,11 +1167,11 @@ void MainScreen::CreateViews() {
 
-#if !PPSSPP_PLATFORM(UWP)
-	// Having an exit button is against UWP guidelines.
+//#if !PPSSPP_PLATFORM(UWP)
+	// Having an exit button is against UWP guidelines?.
```
No, but:
1- Emulators are not allowed anyway in the store (I see the exist button is important to have)
2- Instead of `CoreApplication::Exit()` we are using [TryConsolidateAsync](https://learn.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.applicationview.tryconsolidateasync?view=winrt-22621)

> This method is a programmatic equivalent to a user initiating a close gesture for the app view.

I remember that there was a remarks regarding `CoreApplication::Exit()` but not able to find that anymore.

unknownbrackets replied:
Well, seems reasonable to me then.

-[Unknown]

basharast commented on UI/NativeApp.cpp:
```diff
@@ -491,7 +495,33 @@ void NativeInit(int argc, const char *argv[], const char *savegame_dir, const ch
-
+#elif PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
+	Path memstickDirFile = g_Config.internalDataDirectory / "memstick_dir.txt";
+	auto pathResolved = Path(ResolvePathUWP(memstickDirFile.ToString()));
```
True, I was checking this yesterday and noticed that if I assigned `LocalState` location to `g_Config.internalDataDirectory` would be way better for other places too, I'm getting to understand the project's structure better by time.

basharast commented on Core/PSPLoaders.cpp:
```diff
@@ -418,6 +418,11 @@ bool Load_PSP_ELF_PBP(FileLoader *fileLoader, std::string *error_string) {
 
 	std::string homebrewName = PSP_CoreParameter().fileToStart.ToVisualString();
 	std::size_t lslash = homebrewName.find_last_of("/");
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
```
I was testing this on RetroArch (PSP) version with custom memStick location, and the path was with `\` so I don't know if that happen due to the changes I made.

basharast commented on Core/System.cpp:
```diff
@@ -767,6 +767,9 @@ void InitSysDirectories() {
 	File::CreateDir(GetSysDirectory(DIRECTORY_SAVEDATA));
 	File::CreateDir(GetSysDirectory(DIRECTORY_SAVESTATE));
 	File::CreateDir(GetSysDirectory(DIRECTORY_SYSTEM));
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
```
Some Xbox users asked for this actually, also for `Cheats` folder for easy access, to avoid creating it manually.
if it's unnecessary we can drop this, I don't see it that important, but since I don't use Xbox don't know if it's really needed.

unknownbrackets replied:
Well, I'm fine with it, I just think there's no reason for it to be different for UWP.

I do think cheats should be easier to import...

-[Unknown]

basharast commented on UI/NativeApp.cpp:
```diff
@@ -491,7 +495,33 @@ void NativeInit(int argc, const char *argv[], const char *savegame_dir, const ch
+            }
+			else {
+				ERROR_LOG(SYSTEM, "Couldn't read directory '%s' specified by memstick_dir.txt.", memstickDir.c_str());
+				if (System_GetPropertyBool(SYSPROP_ANDROID_SCOPED_STORAGE)) {
```
I can show question dialog, I totally missed this part of the code, will do a test soon.

basharast commented on UI/MemStickScreen.cpp:
```diff
@@ -124,7 +137,11 @@ MemStickScreen::MemStickScreen(bool initialSetup)
 		}
 	} else {
 		// Detect the current choice, so it's preselected in the UI.
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
+		if (g_Config.memStickDirectory == Path(GetLocalFolder())) {
```
Hmm, I guess I will use `g_Config.internalDataDirectory` which is will be assigned as `LocalState`
What is the deference between `g_extFilesDir` and `internalDataDirectory` usually?
can I just depend on `internalDataDirectory` or this variable is for `Installation` folder?

unknownbrackets replied:
Normally, `g_extFilesDir` is only used by Android currently.  Android has this weird concept of "external" and "internal" storage, neither of which refer to USB or microSD storage.  It's essentially app-private storage.  From a UWP perspective, you could use both to mean `GetLocalFolder()`.

But I think the change to internalDataDirectory is fine too.

-[Unknown]

hrydgard commented on UI/MemStickScreen.cpp:
```diff
@@ -666,7 +711,11 @@ UI::EventReturn ConfirmMemstickMoveScreen::OnConfirm(UI::EventParams &params) {
 				// TODO: Handle failure listing files.
 				std::string error = "Failed to read old directory";
 				INFO_LOG(SYSTEM, "%s", error.c_str());
+#if PPSSPP_PLATFORM(UWP) && !defined(__LIBRETRO__)
+				progressReporter_.Set(error.c_str()); // it's throwing exception with 'iz->T'
```
Good catch, that's pretty bad. Fixing.

basharast committed:
Review's changes, more fixes and better structure

basharast commented on UWP/UWPHelpers/StorageManager.cpp:
```diff
@@ -0,0 +1,384 @@
+	if (result != nullptr && result->Size > 0) {
+		try {
+			auto it = result->First();
+			auto sizeString = it->Current->Value->ToString();
```
I'm not sure if [RetrievePropertiesAsync](https://learn.microsoft.com/en-us/uwp/api/windows.storage.fileproperties.storageitemcontentproperties.retrievepropertiesasync?view=winrt-22621) works perfectly, very little information about it,
based on docs [System.FreeSpace](https://learn.microsoft.com/en-us/windows/win32/properties/props-system-freespace) should return: 

> The amount of free space in a volume, in bytes

when I tried to cast this value (`Platform::Object`) to `int64_t` (`safe_cast<int64_t>(it->Current->Value)`)
it throwing WinRT exception `0x80004002 No such interface supported`,  
so I used `ToString()` instead, I guess this is not correct
it always report "2GB" maybe because of `ToString()`, 
it need someone with better experience to check the conversion

If we can get this to work, it will help for other drives too.

unknownbrackets replied:
Hm, I'm not sure.  A quick search shows people using something more akin to `result["System.FreeSpace"]` in C#/etc.  What if you set a breakpoint or log the FullName of GetType() on the object?

But I guess if we can count on it returning `#(.##)?[PTGMK]B` we can parse it worst case... although the resolution of only "2GB" isn't ideal.

-[Unknown]

basharast replied:
solved, I fetched the result in wrong way, it should be like
```cpp
auto value = result->Lookup(L"System.FreeSpace");
remainingSize = (uint64_t)value;
```

unknownbrackets commented on Common/File/DiskFree.cpp:
```diff
@@ -19,13 +19,25 @@
 #include "Common/Data/Encoding/Utf8.h"
 
+#if PPSSPP_PLATFORM(UWP)
+#include <UWP/UWPHelpers/StorageManager.h>
```
nit: Should use `""` style include.

-[Unknown]

unknownbrackets commented on Common/File/DiskFree.cpp:
```diff
@@ -19,13 +19,25 @@
+	}
+	else
+#endif
+	// Is 'GetDiskFreeSpaceExW' returning wrong values in UWP?
```
No idea.  I mean, it's usable on UWP, but I don't know if it's allowed for all paths.  Would almost seem like a security issue if it "worked" for paths you don't have access to, as it might allow you to probe file existence...

-[Unknown]

basharast committed:
Free space and other fixes

- Fixed freespace value
- added back BOOTABLE case to BROWSE_FOR_FILE
- Fixed fake folder info
- Fixed some code formating

basharast committed:
Update StorageManager.cpp

basharast committed:
better preview path

basharast committed:
Fix _wfopen mode conversion

basharast committed:
Merge branch 'master' into master

basharast committed:
Code cleanup

basharast committed:
Fixed Path issue

basharast committed:
Revert creating ToShortFriendlyPath, more cleanup

Path preview can actually solved within ToVisualString without creating another function, I feel it will cause confusing as both function will be similar.

basharast committed:
Added BrowseFileType::SOUND_EFFECT case

basharast committed:
Merge branch 'master' into master

hrydgard commented:
I'll take another look soon, sorry for the long, long delay!

basharast committed:
Fix UWP build

Replacing NativeMessageReceived with System_PostUIMessage

basharast committed:
Merge branch 'master' of https://github.com/basharast/ppsspp

basharast commented:
> I'll take another look soon, sorry for the long, long delay!

No problem at all, take your time, let me know if there will be any further changes
Many thanks :)

basharast committed:
Merge branch 'master' into master

basharast committed:
Merge branch 'master' of https://github.com/hrydgard/ppsspp

basharast committed:
Resolve launch item and conflicts

- Better handling for launch items
- Resolved code conflicts

basharast committed:
Exclude root choice from memStickScreen (UWP)

This was wrapped before with `#if PPSSPP_PLATFORM(ANDROID)` after it got removed now start to appear in UWP, this should fix it.

hrydgard commented:
Alright, this is looking really good now, there are some nice fixes here.

Just wondering why indentation is now inconsistent betwee different XML files - some get their tabbing changed in this PR, making diffs hard to read (sure, there are workaround but still) and some don't.

With that fixed, I'm happy to merge.

basharast committed:
Correct 'manifest' files formating

basharast commented:
> Just wondering why indentation is now inconsistent betwee different XML files - some get their tabbing changed in this PR, making diffs hard to read (sure, there are workaround but still) and some don't.

Those are 'manifest' files, lines were aligned with spaces instead of tabs, so now will look different
I used VS auto formatting to and it fix and aligned the lines again with tabs instead

<details>
<summary>
Old manifest
</summary>
<img src="https://github.com/hrydgard/ppsspp/assets/3244951/55974451-4d02-4e5e-9723-b985cac7d9ba"/>
</details>

<details>
<summary>
New manifest
</summary>
<img src="https://github.com/hrydgard/ppsspp/assets/3244951/aa23404c-f9f4-4f46-9b53-2195964c02a2"/>
</details>

Hope they are better now after this commit

hrydgard commented:
When looking at the overall PR diff, I still see them with a formatting change. But eh, it's not that important.
